File: benches/priority_queue.rs
use std::collections::BinaryHeap;

#[derive(Default)]
struct MyRand {
    seed: u64,
}

impl MyRand {
    fn next(&mut self) -> u64 {
        self.seed = self.seed.overflowing_mul(123456789).0.overflowing_add(101112131415).0;
        self.seed
    }
}


fn main() {
    let mut rand = MyRand::default();
    let mut pq = BinaryHeap::new();
    for _ in 0..5_000_000 {
        pq.push(rand.next());
        pq.push(rand.next());
        pq.pop();
    }
    println!("{}", pq.len());
}

File: benches/priority_queue.cpp
#include <queue>
#include <iostream>
#include <cstdint>
#include <ranges>

struct MyRand {
    uint64_t seed = 0;

    uint64_t next() {
        seed = seed * 123456789 + 101112131415;
        return seed;
    }
};

using namespace std;

int main() {
    MyRand rand{};
    priority_queue<uint64_t> pq;
    for (int _: views::iota(0, 5'000'000)) {
        pq.push(rand.next());
        pq.push(rand.next());
        pq.pop();
    }
    cout << pq.size() << endl;
}
